### ISSUE TITLE
Copy bins into instrumented directory as-is?

### ISSUE THREAD

**Issue by DenisGorbachev** [wallabyjs/public#198]
https://github.com/fixin-platform/backend-worker/commit/7c812a437f007a35531534439f9a37e4740a5796

Fails on Wallaby with `Test run failure: file bin/worker has invalid syntax`, but runs OK on mocha. Could you please add an option to copy the file unmodified and untouched? (e.g. `passthrough: true`)

---

Note to Googlers: see https://github.com/wallabyjs/public/issues/198#issuecomment-121250969 for solution

**Comment by ArtemGovorov** [comment]
I can't try the code right now, but configuring wallaby [CoffeeScript compiler](http://wallabyjs.com/docs/config/compilers.html) to avoid touching `bin` folder (plus `instrument: false like` like you already have) may work for you:

``` JavaScript
module.exports = function (wallaby) {
  var coffeeCompiler = wallaby.compilers.coffeeScript();
  return {
    files: [
      ...
    ],

    tests: [
      ...
    ],

    compilers: {
      'lib/**/*.coffee': coffeeCompiler,
      'test/**/*.coffee': coffeeCompiler
    }
  };
};
```

(the default value is `**/*.coffee': coffeeCompiler`).

**Comment by ArtemGovorov** [comment]
Removed my previous message, you don't need to configure compilers.
Just place the pattern with `instrument: false` above the broader one (as for each file the first found is applied) and it should work for you.
```
      {pattern: "bin/**/*", instrument: false, load: true, ignore: false},
      "**/*",
```

Please let me know if it works for you.

**Comment by DenisGorbachev** [comment]
Hooray, it works!

I had to prop it up with `fs.chmodSync(path, 0o0755)`, because `bin/worker` was being written out with `-rw-rw-r--`, while I needed `-rwxr-xr-x`. Maybe Wallaby could check if the executable permissions are set on original file, and set it on the copy?

Also, I thought that patterns override each other in ascending order (like variables and config values). Do you think it's a good idea to maintain the abstraction, or would you rather mention the different override order in docs?

**Comment by DenisGorbachev** [comment]
Ah, nope, it doesn't — and the problem is deeper than it seems.

Two problems, actually:

1. `#!/usr/bin/env coffee` gets stripped, so shell tries to run the file as if it were a bash script.
1. bin/worker `require`s other files, which are instrumented by Wallaby, hence `ReferenceError: $_$wp is not defined`

Does that mean the bin tests have to be confined to mocha?

**Comment by DenisGorbachev** [comment]
also, bin/workers `requires`s node_modules, which have to be copied at startup, which is a time-consuming process.

**Comment by DenisGorbachev** [comment]
I've just set up running original bin/worker right from the project directory. Lost the code coverage, but at least the tests work.

https://github.com/fixin-platform/backend-worker/commit/9fbf0e29e24d9d465a97b523f0a48adca2c3c44b

Note: `process.env.ROOT_DIR` is set manually through `wallaby.js`.

**Comment by ArtemGovorov** [comment]
Ok, now I had an opportunity to look int the code a bit deeper.

> I thought that patterns override each other in ascending order (like variables and config values)

Pattern checking logic was implemented the same way as in Karma config to make it easier for existing Karma users to re-use their config for wallaby. I have added a note about it to the [file object docs (pattern section)](http://wallabyjs.com/docs/config/files.html#file-object).

> various bin/worker issues, lost coverage

Wallaby.js doesn't support cross-process coverage yet (I don't think any of the existing cove coverage solution do support it). We have some future plans to support collecting coverage from `fork`-ed workers. But when you `spawn` a new process (as an executable), wallaby has no way to plug in its infrastructure to collect code coverage, to tell the code where to look for local node modules, etc.

Your solution of spawning the original file, as opposed to copying it, is the best way to make it work. It also plays nicely with the (correct IMHO) way you test - you have some tests to test start and communication system with the worker, then you have some separate tests for `WorkerFactory`. So you test separate things things separately and are getting all the coverage for your code anyway (except for the worker code itself).

> I had to prop it up with `fs.chmodSync(path, 0o0755)`

I know you have already removed it, but I thought I'd still mention that a better place for stuff like that would be in wallaby.js `bootstrap` function rather than in tests, to keep tests test runner agnostic.
